### DISCUSSION TITLE
Cannot boot Windows10(1809) on targetcli(LIO)

### DISCUSSION THREAD

**Discussion by WalonLi** [ipxe/ipxe#249] [Q&A] [answered]
Hi experts,

I got an issue that system cannot boot windows iSCSI target via iPXE, but installation is always good. See my below cases.
**Case 1 - targetcli**
1.  Install Windows10(1809) on iSCSI target. => PASS.
2.  Boot Windows10 on iSCSI target => FAIL, Windows Boot Manger shows "\windows\system32\ntoskrnl.exe" can't be loaded, status code is 0xc0000001.
![1](https://user-images.githubusercontent.com/8321049/106858284-652a6100-66fc-11eb-8cc1-fbdb91be4736.png)

**Case 2 - iscsitarget (IET)**
1.  Install Windows10(1809) on iSCSI target. => PASS.
2.  Boot Windows10 on iSCSI target => PASS.

Do you have any idea about that? System seems already boot to windows boot manager but the file can not be loaded. Only difference between two cases is iSCSI server (targetcli vs iscsitarget), others are same(iPXE script, Windows and WinPE). 

Also, have some more test for you reference.
1. Build the latest iPXE on github => Still got failure.
2. Had a try with edk2(UEFI) iSCSI initiator, everything works fine.
3. If target OS is linux (Redhat/SLES), everything works fine.
4. Checked dmesg on target server, it looks like LIO throws a lot of FILEIO ERROR with case1. Not sure it's helpful.
![2](https://user-images.githubusercontent.com/8321049/106858534-c8b48e80-66fc-11eb-85ff-9323913b1cd6.png)

**Boot Script**
```
#!ipxe
set root-path iscsi:192.168.112.1::::iqn.2020-04.test.abc:disk1
set initiator-iqn iqn.2020-04.abc.init:init
dhcp
set gateway 0.0.0.0
set keep-san 1
sanboot --filename \EFI\BOOT\BootX64.efi ${root-path}
boot
```

Thanks!

**Comment by mcb30** [comment, marked as answer]
This looks to be a known bug in older versions of the LIO target: see e.g. https://lists.debian.org/debian-kernel/2016/01/msg00632.html

Upgrading your target software (including the kernel, since the relevant code is implemented within the kernel module) should fix the problem.

**Reply by WalonLi** [reply]
I had checked this page. I didn't see the [linux-kernel](https://github.com/torvalds/linux/blob/master/drivers/target/target_core_file.h ) with this patch on github. So it might linux-kernel doesn't pick this change from debian? I'm not familiar on that, just a guess. 
Actually, I tried two distro, and got similar error. I/O size 9696256 is bigger than that.
![1](https://user-images.githubusercontent.com/8321049/106913243-beb37f80-673e-11eb-8365-ba816f8032dc.png)
![2](https://user-images.githubusercontent.com/8321049/106913294-cffc8c00-673e-11eb-8bf7-c665a365761e.png)

By the way, may I know what's your distro or kernel version when you were developing this feature? Appreciate your help!

**Reply by mcb30** [reply]
> I had checked this page. I didn't see the [linux-kernel](https://github.com/torvalds/linux/blob/master/drivers/target/target_core_file.h) with this patch on github. So it might linux-kernel doesn't pick this change from debian? I'm not familiar on that, just a guess.

The message at https://lists.debian.org/debian-kernel/2016/01/msg00632.html states that the issue "doesn't affect upstream anymore, because they have rewritten LIO to use vfs_iter_write".  I'm guessing that rewrite was probably commit https://github.com/torvalds/linux/commit/6b9a44d0939ca6235c72b811bd55b462d6a0a553

> Actually, I tried two distro, and got similar error. I/O size 9696256 is bigger than that.
> ![1](https://user-images.githubusercontent.com/8321049/106913243-beb37f80-673e-11eb-8365-ba816f8032dc.png)

The kernel versions shown in your picture (3.10.0-1062.el7 and 4.4.0-87) are both at least four years old.

> By the way, may I know what's your distro or kernel version when you were developing this feature? Appreciate your help!

I tend to use Fedora, though I have no idea which version I would have been using at the time of developing, sorry.

This is definitely not an iPXE issue: the problem is purely in the interaction between your Windows iSCSI initiator and your iSCSI target.  You may want to try raising the issue at http://linux-iscsi.org/wiki/Support instead.

Good luck!

**Reply by WalonLi** [reply]
OK, sounds I should raise the issue to linux.org. Update kernel (5.10.13-1.el7.elrepo.x86_64) on centos is not work as well.
Anyway, thanks a lot.
